sub-packages/bionemo-llm/src/bionemo/llm/model/loss.py

```diff
@@ -174,43 +173,42 @@ def forward(
         if "labels" not in batch:
             raise ValueError("Labels not provided in the batch. These are required for this loss computation.")
 
-        # NOTE: token_logits is [sequence, batch] but labels and other fiels, including the loss are [batch, sequence]
+        # NOTE: token_logits is [sequence, batch] but labels and other fields, including the loss are [batch, sequence]
         unreduced_token_loss = unreduced_token_loss_fn(forward_out["token_logits"], batch["labels"])  # [b s]
 
         # TODO(@jstjohn) also handle different output keys, like the sequence loss.
 
-        # Compute loss over "valid" tokens in the microbatch, i.e. the non-masked tokens.
-        # The loss is not normalized, so you need to divide by the number of non-masked
-        # tokens (loss_mask.sum()) to compute the mean loss per token.
-        loss_for_microbatch, num_valid_tokens_in_microbatch = masked_token_loss(
+        # NOTE(SKH) This now computes the sum of all losses over the group as well as the number of valid tokens in the microbatch.
+        group_token_loss_sum, num_valid_tokens_in_microbatch = masked_token_loss(
             unreduced_token_loss, batch["loss_mask"]
         )
 
-        # Get the context parallel size for some normalizations and reductions.
-        cp_size = parallel_state.get_context_parallel_world_size()
+        # We can compute the loss for the group by dividing the summed loss by the number of valid tokens in the group.
+        loss_for_microbatch = group_token_loss_sum / num_valid_tokens_in_microbatch
 
         # If we do not drop the last partial batch of validation, we need to do fancy reduction handling to support
         #  reducing the loss across the data parallel group.
         if self.validation_step and not self.val_drop_last:
-            if loss_for_microbatch.isnan():
```

Comment:
I don't think you need this anymore with the new formulation. Follow the pattern of https://github.com/NVIDIA/NeMo/blob/main/nemo/lightning/megatron_parallel.py#L1733-L1788 which should work for either the drop last and regular case? I think that's part of the appeal.